src/Models/model.jl

```diff
@@ -0,0 +1,87 @@
+# pros:
+# easy tests
+# easy inspection
+# high flexibility
+
+# cons:
+# very functional
+
+abstract type AbstractModel end
+abstract type AbstractEquation end
+abstract type AbstractBoundaryCondition end
+abstract type Rate{N} end
+abstract type SteppingStyle end
+abstract type AbstractDirection end
+abstract type AbstractTendency end
+abstract type AbstractConstraint end
+
+# example of a concrete model
+struct MyModel{EQ <: AbstractEquation, CO <: AbstractConstraint} <: AbstractModel
+    eq1::EQ
+    eq2::EQ
+    eq3::EQ
+    co1::CO
+    co2::CO
+end
+
+# example of a concrete equation
+struct Density{B <: AbstractBoundaryCondition, T <: AbstractTendency, C <: AbstractConstraint} <: AbstractEquation
+    top::B
+    btm::B
+    advection::T
+    diffusion::T
+    positivity::T
+end
+
+# example of a concrete boundary condition
+struct LatentHeatFlux <: AbstractBoundaryCondition 
+    bc_state # e.g., prescribed T_sfc
+end
+
+# example of concrete stepping styles
+struct Explicit{R <: Rate{N}} <: SteppingStyle end
+struct Implicit{R <: Rate{N}} <: SteppingStyle end
+
+# example of concrete concrete directions
+struct HorizontalDirection <: AbstractDirection end
+struct VerticalDirection <: AbstractDirection end
+
+# examples of a concrete tendencies
+struct Advection{S <: SteppingStyle, D <: AbstractDirection} <: AbstractTendency end
+struct Diffusion{S <: SteppingStyle, D <: AbstractDirection} <: AbstractTendency end
+struct Coriolis{FT, S <: SteppingStyle} <: AbstractTendency 
+    rotation_rate::FT
+end
+
+# examples of a concrete constraints types
+struct Positivity{R <: Rate} <: AbstractConstraint end
+struct ThermodynamicEquilibrium{R <: Rate} <: AbstractConstraint end
+
+# returns all equations
+return_equations(model::AbstractModel)
```

Comment:
It might be helpful to see an example here. For `MyModel `, would this be:
```julia
return_equations(m::MyModel) = (m.eq1, m.eq2, m.eq3)
```
?
Also, where and how will this method be used?

Reply:
Yes. I am setting up and example.

Reply:
But that you can guess the implementation is a good sign :)